src/io.c

```diff
@@ -72,76 +66,40 @@ uint32_t memory_read_str(memory_t *m, uint8_t *dst, uint32_t addr, uint32_t max)
     return len + 1;
 }
 
-uint32_t memory_ifetch(memory_t *m, uint32_t addr)
-{
-    const uint32_t addr_lo = addr & mask_lo;
-    assert((addr_lo & 1) == 0);
-
-    chunk_t *c = m->chunks[addr >> 16];
-    assert(c);
-    return *(const uint32_t *) (c->data + addr_lo);
-}
-
-uint32_t memory_read_w(memory_t *m, uint32_t addr)
+uint32_t memory_ifetch(uint32_t addr)
 {
-    const uint32_t addr_lo = addr & mask_lo;
-    chunk_t *c = m->chunks[addr >> 16];
-    return *(const uint32_t *) (c->data + addr_lo);
+    return *(const uint32_t *) (data_memory_base + addr);
 }
 
-uint16_t memory_read_s(memory_t *m, uint32_t addr)
-{
-    const uint32_t addr_lo = addr & mask_lo;
-    chunk_t *c = m->chunks[addr >> 16];
-    return *(const uint16_t *) (c->data + addr_lo);
-}
+#define MEM_READ_IMPL(size, type)                   \
+    type memory_read_##size(uint32_t addr)          \
+    {                                               \
+        return *(type *) (data_memory_base + addr); \
+    }
 
-uint8_t memory_read_b(memory_t *m, uint32_t addr)
-{
-    chunk_t *c = m->chunks[addr >> 16];
-    return c->data[addr & mask_lo];
-}
+MEM_READ_IMPL(w, uint32_t);
+MEM_READ_IMPL(s, uint16_t);
+MEM_READ_IMPL(b, uint8_t);
 
-void memory_write(memory_t *m, uint32_t addr, const uint8_t *src, uint32_t size)
+void memory_write(memory_t *mem,
```

Comment:
Ditto. Define it as a static inline function.

Reply:
Ditto.